File: 0029-Divide Two Integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        positive = ((dividend < 0) == (divisor < 0))
        dividend, divisor = abs(dividend), abs(divisor)
        res = 0
        while dividend >= divisor:
            temp, i = divisor, 1
            while dividend >= temp:
                res += i
                dividend -= temp
                i <<= 1
                temp <<= 1
        if not positive:
            res = -res
        return min(max(res, -2**31), 2**31-1)

File: 0029-Divide Two Integers.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        long res = 0;
        bool positive = ((dividend < 0) == (divisor < 0));
        
        long dividend2 = labs(dividend);
        long divisor2 = labs(divisor);
        while(dividend2 >= divisor2){
            long temp = divisor2;
            long i = 1;
            // 指加速，不然溢出
            while(dividend2 >= temp){
                res += i;
                dividend2 -= temp;
                temp = temp << 1;
                i= i << 1;
            }
            
        }
        if(!positive){
            res = -res;
        }
        if(res < INT_MIN){
            res = INT_MIN;
        }
        if(res > INT_MAX){
            res = INT_MAX;
        }
        
        return res;
    }
};